src/main/scala/latis/service/hapi/InfoService.scala

```diff
@@ -17,36 +18,31 @@ class InfoService[F[_]: Concurrent](alg: InfoAlgebra[F]) extends Http4sDsl[F] {
   val service: HttpRoutes[F] =
     HttpRoutes.of[F] {
       case GET -> Root / "info"
-          :? DatasetMatcher(ds)
-          +& IdMatcher(id)
-          +& ParamMatcher(params) =>
-        // Allow both "dataset" or "id" for backwards compatibility
-        Either.fromOption(ds <+> id, Status.`1400`).fold(
-          err => {
-            logger.info(err.message)
-            InternalServerError(HapiError(Status.`1400`).asJson)
-          },
-          dataset => {
-            val ps = params.map(_.distinct)
-            alg.getMetadata(dataset, ps).leftMap {
-              case UnknownId(_)        => Status.`1406`
-              case UnknownParam(_)     => Status.`1407`
-              case err@MetadataError(_)      => logger.info(err.toString); Status.`1501`
-              case err@UnsupportedDataset(_) => logger.info(err.toString); Status.`1501`
-            }.fold(
-              err => err match {
-                case Status.`1406` | Status.`1407` =>
-                  logger.info(err.message)
-                  NotFound(HapiError(err).asJson)
-                case _ =>
-                  logger.info(err.message)
-                  InternalServerError(HapiError(err).asJson)
-              },
-              res => Ok(
-                InfoResponse(HapiService.version, Status.`1200`, res).asJson
-              )
-            ).flatten
+        :? DatasetMatcher(ds)
+        +& IdMatcher(id)
+        +& ParamMatcher(params) =>
+        (for {
+          // Allow both "dataset" or "id" for backwards compatibility
+          dataset  <- EitherT.fromOption[F](ds <+> id, Status.`1400`)
+          ps        = params.map(_.distinct)
+          metadata <- alg.getMetadata(dataset, ps).leftMap {
+            case UnknownId(_)        => Status.`1406`
+            case UnknownParam(_)     => Status.`1407`
+            case err@MetadataError(_)      => logger.info(err.toString); Status.`1501`
+            case err@UnsupportedDataset(_) => logger.info(err.toString); Status.`1501`
           }
+        } yield metadata).foldF(
+          err => err match {
+            case Status.`1406` | Status.`1407` =>
+              logger.info(err.message)
+              NotFound(HapiError(err).asJson)
+            case _ =>
+              logger.info(err.message)
+              BadRequest(HapiError(err).asJson)
```

Comment:
We want the 1400 to correspond to a 400 but we want the 1501s (and anything we didn't explicitly handle) to correspond to a 500.

Reply:
Do you mean like this?

```scala
          err => err match {
            case Status.`1400` =>
              logger.info(err.message)
              BadRequest(HapiError(err).asJson)
            case Status.`1406` | Status.`1407` =>
              logger.info(err.message)
              NotFound(HapiError(err).asJson)
            case _ =>
              logger.info(err.message)
              InternalServerError(HapiError(err).asJson)
          }
```

Reply:
Yep.